### PR TITLE
Moving AllowAnonymous attribute to Authorization

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#3246 
@rynowak @Eilon 

Corresponding Security PR https://github.com/aspnet/Security/pull/520

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs [hunk 1]
```diff
@@ -53,6 +58,11 @@ public void OnProvidersExecuting(ApplicationModelProviderContext context)
                     {
                         actionModel.Filters.Add(new AuthorizeFilter(policy));
                     }
+
+                    foreach (var attribute in actionModel.Attributes.OfType<AllowAnonymousAttribute>())
+                    {
+                        actionModel.Filters.Add(new AllowAnonymousFilter());
+                    }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: unit tests for this provider?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: Added.

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs [hunk 2]
```diff
@@ -44,6 +44,11 @@ public void OnProvidersExecuting(ApplicationModelProviderContext context)
                     controllerModel.Filters.Add(new AuthorizeFilter(policy));
                 }
 
+                foreach (var attribute in controllerModel.Attributes.OfType<AllowAnonymousAttribute>())
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: Should this be checking the concrete attribute, or should this be checking for any attribute that implements the `Microsoft.AspNet.Authorization.IAllowAnonymous` interface?

Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: (Otherwise what's the point of that interface?)

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: We can do that. In that case, should this https://github.com/aspnet/Mvc/pull/3330/files#diff-989bc8123071d2e7f36cc5d32dc275ffL41 also be changed to use its interface `IAuthorizeData`?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/AuthorizationApplicationModelProvider.cs]: Yes, that's the same kind of issue :+1:

**Comment by ajaybhargavb** [comment]
Updated. Note: https://github.com/aspnet/Security/pull/520/files#diff-091a4b1a0da83c57ebac3d8e0e7f1c8dR60 needed change for this to work.

**Comment by Eilon** [comment]
:shipit:

**Pull request merged**